File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:

    void findCandidates(vector<int>& candidates,int target,set<vector<int>,less<>>& unique,vector<int>& temp, int i) {
        if(i==candidates.size()) {
            if(target==0) {
                unique.insert(temp);
            }
            return;
        }
        if(candidates[i]<=target) {
            temp.push_back(candidates[i]);
            findCandidates(candidates,target-candidates[i],unique,temp,i);
            temp.pop_back();
        }
        findCandidates(candidates,target,unique,temp,i+1);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        set<vector<int>,less<>> unique;
        vector<int> temp;
        findCandidates(candidates,target,unique,temp,0);
        vector<vector<int>> result;
        for(auto it : unique) {
            result.push_back(it);
        }
        return result;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void backtrack(int[] candidates, int target, int start, List<Integer> temp, List<List<Integer>> result) {
        if(start==candidates.length) {
            if(target==0) result.add(new ArrayList<>(temp));
            return;
        }
        if(target>=candidates[start]) {
            temp.add(candidates[start]);
            backtrack(candidates,target-candidates[start],start,temp,result);
            temp.remove(temp.size()-1);
        }
        backtrack(candidates,target,start+1,temp,result);
        
        // if(target == 0) {
        //     result.add(new ArrayList<>(temp));
        //     return;
        // }
        // for(int i=start;i<candidates.length;i++) {
        //     if(target>=candidates[i]) {
        //         temp.add(candidates[i]);
        //         backtrack(candidates,target-candidates[i],i,temp,result);
        //         temp.remove(temp.size()-1);
        //     }
        // }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates,target,0,new ArrayList<>(),result);
        return result;
    }
}